### PR TITLE
Plugin refactoring #28 #32 #39 #37

### LINKED ISSUE #28: Refactor this code
I don't feel like opening three issues for this code, but this really needs to be refactored:
- There is copy & paste all over the code (it's actually most of it within and across all files related to collection of data).
- There is no separation between snap interaction & metrics collection. In such simple case we may live without separate package but at leas within main one functions should deal only with collection or only with namespaces or only with snap api.
- Functions that can return error, should use this opportunity instead of panic (unless it's really justified, but I don't see the case in this plugin). So please remove `handleErr`.

@intelsdi-x/plugin-maintainers

### PR DESCRIPTION
**Pull request by sandlbn**
Fixes #28 #32 #39 #37 

@intelsdi-x/plugin-maintainers 

Summary of changes:
-Moved libvirt functions to separate package
-Removed dependency on etree
-Moved libvirt to libvirtcollector 

Testing done:
- Unit Tests
- Openstack Cluster

### CODE REVIEW

**Comment by katarzyna-z** [comment]
consider update of godeps or replacing godep with glide, snap commit in Godep.json is quite old.

File: libvirtcollector/libvirtcollector.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	// Type of plugin
+	Type = plugin.CollectorPluginType
+	// Vendor  prefix
+	Vendor = "intel"
+	// Plugin plugin name
+	Plugin = "libvirt"
+	// Version of plugin
+	Version = 10
+	// each metric starts with prefix "/intel/libvirt/<domain_id>"
+	lengthOfNsPrefix = 3
```
Comment (katarzyna-z) [inline libvirtcollector/libvirtcollector.go]: lengthOfNsPrefix seems to be unused.

File: libvirtcollector/libvirtcollector.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+		if nova.(bool) {
+			meta, err = wrapper.GetNovaMetadata(id)
+			if err != nil {
+				return metrics, err
+			}
+		}
+
+		for _, mt := range filterNamespace("network", mts) {
+			ns := copyNamespace(mt)
+			ns[2].Value, err = id.GetName()
```
Comment (katarzyna-z) [inline libvirtcollector/libvirtcollector.go]: It is just suggestion but consider creation of const  variables with descriptive names for parts of namespace to replace this `2` with  const variable. I think about something like `ns[nsNamePosition].Value` . Below there are similar cases for `ns[4]`, `n[5]`.

File: libvirtcollector/libvirtcollector.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+var defaultURI = "qemu:///system"
+
+// NewLibvirtCollector returns new instance of collector
+func NewLibvirtCollector() *LibvirtCollector {
+	return &LibvirtCollector{}
+
+}
+
+// CollectMetrics returns collected metrics
+func (p *LibvirtCollector) CollectMetrics(mts []plugin.MetricType) ([]plugin.MetricType, error) {
```
Comment (katarzyna-z) [inline libvirtcollector/libvirtcollector.go]: It is up to you but consider splitting of CollectMetrics into functions if it is possible, now it long method.

Comment (sandlbn) [inline libvirtcollector/libvirtcollector.go]: @katarzyna-z I was thinking about it but code is so simple there, that I think is not necessary.

File: libvirtcollector/libvirtcollector.go (hunk 2 above)
Comment (sandlbn) [inline libvirtcollector/libvirtcollector.go]: Done

**Comment by kindermoumoute** [comment]
@sandlbn, as you refactor the plugin, why not using the [new Go lib](https://github.com/intelsdi-x/snap-plugin-lib-go/tree/master/v1/plugin)?
(switching to the new lib is pretty quick)

**Comment by lmroz** [comment]
@sandlbn Code looks much cleaner now :+1: I hope I'll have some time to have deeper look, but I picked up one thing, that might just need clarification:

In `CollectMetrics` you are doing query to libvirt for each metric, and this query (at least in some cases) picks up all stats for each path/interface, so perhaps can be done at most once in this function. I don't know how "heavy" this call is, but potentially it's not only overkill but also performance issue.

**Comment by sandlbn** [comment]
@kindermoumoute Done ! Thanks for hint

**Comment by sandlbn** [comment]
@lmroz it's very light call, but I agree not need to be done 6 times, per collection.

File: libvirtcollector/libvirtcollector.go (hunk 3 above)
Comment (katarzyna-z) [inline libvirtcollector/libvirtcollector.go]: OK

**Comment by sandlbn** [comment]
@lmroz ok done. now it's only one call per collection.

File: libvirt/domain.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+func GetNovaMetadata(domain libvirt.VirDomain) (map[string]string, error) {
+	var novaMetadata map[string]string
+	domXML, err := getDomainXML(domain)
+	if err != nil {
+		return novaMetadata, err
+	}
+
+	if domXML.Metadata.NovaInstance != nil {
+
+		novaMetadata["nova_flavor"] = domXML.Metadata.NovaInstance.NovaFlavor.AttrName
```
Comment (katarzyna-z) [inline libvirt/domain.go]: at this point `novaMetadata` is `nil `, here there are assignments to `nil`, see this simple example:

```go
package main

func main() {

	var novaMetadata map[string]string
	novaMetadata["test"] = "test-string"
}
```

Comment (sandlbn) [inline libvirt/domain.go]: @katarzyna-z sorry I fixed that on my branch but didn't pushed.

File: libvirt/domain.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+//GetDomainInterfaces returns slice of available network devices
+func GetDomainInterfaces(domain libvirt.VirDomain) ([]string, error) {
+	var interfaces []string
+	domXML, err := getDomainXML(domain)
+	if err != nil {
+		return interfaces, err
+	}
+	if domXML.Devices.Interface == nil || domXML.Devices.Interface.Target == nil {
```
Comment (katarzyna-z) [inline libvirt/domain.go]: It seems that you should check if `domXML.Devices` is not `nil` before this line.

File: libvirt/domain.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+		novaMetadata["nova_disk_mb"] = domXML.Metadata.NovaInstance.NovaFlavor.NovaDisk.Text
+		novaMetadata["nova_ephemeral_mb"] = domXML.Metadata.NovaInstance.NovaFlavor.NovaEphemeral.Text
+		novaMetadata["nova_memory_mb"] = domXML.Metadata.NovaInstance.NovaFlavor.NovaMemory.Text
+		novaMetadata["nova_swap_mb"] = domXML.Metadata.NovaInstance.NovaFlavor.NovaSwap.Text
+		novaMetadata["nova_vcpus"] = domXML.Metadata.NovaInstance.NovaFlavor.NovaVcpus.Text
+		novaMetadata["nova_name"] = domXML.Metadata.NovaInstance.NovaName.Text
+		novaMetadata["nova_owner"] = domXML.Metadata.NovaInstance.NovaOwner.NovaUser.AttrUUID
+		novaMetadata["nova_package"] = domXML.Metadata.NovaInstance.NovaPackage.AttrVersion
+		novaMetadata["nova_uuid"] = domXML.UUID.Text
+
```
Comment (sandlbn) [inline libvirt/domain.go]: ok,Done.

File: example/task-example.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+         /intel/libvirt/*/network/*/rxerrs: {}
+         /intel/libvirt/*/network/*/rxpackets: {}
+         /intel/libvirt/*/network/*/txbytes: {}
+         /intel/libvirt/*/network/*/txdrop: {}
+         /intel/libvirt/*/network/*/txerrs: {}
+         /intel/libvirt/*/network/*/txpackets: {}
+      publish:
+        - plugin_name: "file"
+          config:
+            file: /tmp/libvirt_metrics.log"
```
Comment (katarzyna-z) [inline example/task-example.yaml]: Missing `"` before `/tmp/libvirt_metrics.log`. Please add a new line at the end of this file.

File: example/task-example.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+         /intel/libvirt/*/memory/max: {}
+         /intel/libvirt/*/memory/mem: {}
+         /intel/libvirt/*/memory/min_fault: {}
+         /intel/libvirt/*/memory/nr: {}
+         /intel/libvirt/*/memory/rss: {}
+         /intel/libvirt/*/memory/swap_in: {}
+         /intel/libvirt/*/memory/swap_out: {}
+         /intel/libvirt/*/memory/unused: {}
+         /intel/libvirt/*/network/*/rxbytes: {}
+         /intel/libvirt/*/network/*/rxdrop: {},
```
Comment (katarzyna-z) [inline example/task-example.yaml]: Remove this comma

**Comment by katarzyna-z** [comment]
Please squash your commits.

**Comment by katarzyna-z** [comment]
After some time I see following error in snap logs:
```
DEBU[2016-11-14T14:07:25+01:00] plugin selected                               _module=control-routing block=select hitcount=339 index=collector:libvirt:v10:id1 pool size=1 strategy=least-recently-used
DEBU[2016-11-14T14:07:25+01:00] collector run completed                       _module=scheduler-job block=run job-type=collector metric-count=0
ERRO[2016-11-14T14:07:25+01:00] collector run error                           _module=scheduler-job block=run error=rpc error: code = 2 desc = [Code-38] [Domain-0] Failed to find user record for uid '1000': Too many open files job-type=collector
WARN[2016-11-14T14:07:25+01:00] Task failed                                   _block=spin _module=scheduler-task consecutive failure limit=10 consecutive failures=10 error=rpc error: code = 2 desc = [Code-38] [Domain-0] Failed to find user record for uid '1000': Too many open files task-id=4873a3c7-7f5b-4cba-b043-3a373a4bf425 task-name=Task-4873a3c7-7f5b-4cba-b043-3a373a4bf425
ERRO[2016-11-14T14:07:25+01:00] Task disabled due to consecutive failures     _block=spin _module=scheduler-task consecutive failures=10 error=rpc error: code = 2 desc = [Code-38] [Domain-0] Failed to find user record for uid '1000': Too many open files task-id=4873a3c7-7f5b-4cba-b043-3a373a4bf425 task-name=Task-4873a3c7-7f5b-4cba-b043-3a373a4bf425
DEBU[2016-11-14T14:07:25+01:00] event received                                _block=handle-events _module=scheduler-events disabled-reason=Task disabled with error: rpc error: code = 2 desc = [Code-38] [Domain-0] Failed to find user record for uid '1000': Too many open files event-namespace=Scheduler.TaskDisabled task-id=4873a3c7-7f5b-4cba-b043-3a373a4bf425
DEBU[2016-11-14T14:07:25+01:00] plugin unsubscription                         _block=subscriptionGroup.unsubscribePlugins _module=control name=libvirt type=collector version=10
DEBU[2016-11-14T14:07:25+01:00] plugin unsubscription                         _block=subscriptionGroup.unsubscribePlugins _module=control name=file type=publisher version=-1
DEBU[2016-11-14T14:07:25+01:00] handling plugin unsubscription event          _block=subscribe-pool _module=control-runner event=Control.PluginUnsubscribed plugin-name=file plugin-type=publisher plugin-version=-1
```
checking task status using snapctl:
```
$ snapctl task list
ID 					 NAME 						 STATE 		 HIT 	 MISS 	 FAIL 	 CREATED 		 LAST FAILURE
4873a3c7-7f5b-4cba-b043-3a373a4bf425 	 Task-4873a3c7-7f5b-4cba-b043-3a373a4bf425 	 Disabled 	 349 	 0 	 10 	 2:01PM 11-14-2016 	 rpc error: code = 2 desc = [Code-38] [Domain-0] Failed...
```
I tested with this task manifest:
```

---
  version: 1
  schedule:
    type: "simple"
    interval: "1s"
  max-failures: 10
  workflow:
    collect:
      metrics:
         /intel/libvirt/*/cpu/cputime: {}
         /intel/libvirt/*/disk/*/rdbytes: {}
         /intel/libvirt/*/disk/*/rdreq: {}
         /intel/libvirt/*/disk/*/wrbytes: {}
         /intel/libvirt/*/disk/*/wrreq: {}
         /intel/libvirt/*/memory/actual_balloon: {}
         /intel/libvirt/*/memory/available: {}
         /intel/libvirt/*/memory/major_fault: {}
         /intel/libvirt/*/memory/max: {}
         /intel/libvirt/*/memory/mem: {}
         /intel/libvirt/*/memory/min_fault: {}
         /intel/libvirt/*/memory/nr: {}
         /intel/libvirt/*/memory/rss: {}
         /intel/libvirt/*/memory/swap_in: {}
         /intel/libvirt/*/memory/swap_out: {}
         /intel/libvirt/*/memory/unused: {}
         /intel/libvirt/*/network/*/rxbytes: {}
         /intel/libvirt/*/network/*/rxdrop: {}
         /intel/libvirt/*/network/*/rxerrs: {}
         /intel/libvirt/*/network/*/rxpackets: {}
         /intel/libvirt/*/network/*/txbytes: {}
         /intel/libvirt/*/network/*/txdrop: {}
         /intel/libvirt/*/network/*/txerrs: {}
         /intel/libvirt/*/network/*/txpackets: {}
      publish:
        - plugin_name: "file"
          config:
            file: "/tmp/libvirt_metrics.log"
```
Are you able to reproduce this issue?

**Comment by dgarth** [comment]
I can confirm that this new version fixes #32 #37 and #39 in my environment. #38 persists.
I can also confirm the error above. It happens after the same amount of Task-Hits (using the *InfluxDB* publisher instead of *File*).

There seems to be a new issue with the cputime metric: The tag to get per-processor/core data is missing for me. The only tags are "domain_id", "source" and the OpenStack related ones.

**Comment by sandlbn** [comment]
@katarzyna-z yes, I'm debugging now, it's really strange. Which version of go are you using ?

**Comment by katarzyna-z** [comment]
@sandlbn `go version go1.6.2 linux/amd64`

**Comment by sandlbn** [comment]
@katarzyna-z ok, I fixed connection closing problem.

**Comment by sandlbn** [comment]
@dgarth yes I just added vcpu tagging, please check that you are using this feature with kvm. I tested both with kvm and without and without kvm libvirt will return error.

**Comment by sandlbn** [comment]
@dgarth memory reporting is also fixed. On qemu /intel/libvirt/web_devel/memory/actual_balloon and /intel/libvirt/web_devel/memory/rss are only available.

**Comment by katarzyna-z** [comment]
LGTM

**Pull request merged**